### ISSUE TITLE
Differentiating the forward-mode (JVP) rule of `linalg.det` / `slogdet` / `linalg.solve` silently returns a wrong second derivative

### ISSUE THREAD

**Issue by shadow-orange41** [pytorch/pytorch#192540]
### 🐛 Describe the bug

For some operations (`linalg.det`, `linalg.slogdet`, `linalg.solve`), a second-order
derivative taken with `torch.func` is silently wrong. 

The APIs affected are:
- `torch.func.jacfwd(torch.func.jacfwd(f))`
- `torch.func.jacrev(torch.func.jacfwd(f))`
- `torch.func.jvp(torch.func.jvp(f))`
- `torch.func.grad(torch.func.jvp(f))`

The operators affected are `torch.linalg.det`, `torch.linalg.slogdet` (and `torch.logdet`),
and `torch.linalg.solve`.

The first-order derivative is correct in every mode.

```python
import torch

A = torch.tensor([[1.3, 0.4], [-0.7, 0.9]], dtype=torch.float64)
f = torch.linalg.det

# For a 2x2, d^2 det(A) / dA dA is the constant tensor H[i,j,k,l] = eps_ik * eps_jl.
H_true = torch.zeros(2, 2, 2, 2, dtype=torch.float64)
H_true[0, 0, 1, 1] = H_true[1, 1, 0, 0] = 1.0
H_true[0, 1, 1, 0] = H_true[1, 0, 0, 1] = -1.0

BASIS = [(i, j) for i in range(2) for j in range(2)]


def unit(i, j):
    E = torch.zeros(2, 2, dtype=torch.float64)
    E[i, j] = 1.0
    return E


def jvp_f(V):
    return lambda X: torch.func.jvp(f, (X,), (V,))[1]


def hessian_via_jvp_jvp():
    H = torch.empty(2, 2, 2, 2, dtype=torch.float64)
    for i, j in BASIS:
        for k, l in BASIS:
            H[i, j, k, l] = torch.func.jvp(jvp_f(unit(i, j)), (A,), (unit(k, l),))[1]
    return H


def hessian_via_grad_jvp():
    H = torch.empty(2, 2, 2, 2, dtype=torch.float64)
    for i, j in BASIS:
        H[i, j] = torch.func.grad(jvp_f(unit(i, j)))(A)
    return H


fwd, rev = torch.func.jacfwd, torch.func.jacrev
for name, H in [
    ("jacfwd(jacfwd)   inner fwd", fwd(fwd(f))(A)),
    ("jacrev(jacfwd)   inner fwd", rev(fwd(f))(A)),
    ("jvp(jvp)         inner fwd", hessian_via_jvp_jvp()),
    ("grad(jvp)        inner fwd", hessian_via_grad_jvp()),
    ("jacfwd(jacrev)   inner rev", fwd(rev(f))(A)),
    ("jacrev(jacrev)   inner rev", rev(rev(f))(A)),
]:
    print(f"  {name}:  max|H - H_true| = {(H - H_true).abs().max():.3e}")

```

Output:

```
  jacfwd(jacfwd)   inner fwd:  max|H - H_true| = 1.166e+00
  jacrev(jacfwd)   inner fwd:  max|H - H_true| = 1.166e+00
  jvp(jvp)         inner fwd:  max|H - H_true| = 1.166e+00
  grad(jvp)        inner fwd:  max|H - H_true| = 1.166e+00
  jacfwd(jacrev)   inner rev:  max|H - H_true| = 8.882e-16
  jacrev(jacrev)   inner rev:  max|H - H_true| = 1.110e-15
```

### Versions

Collecting environment information...
PyTorch version: 2.14.0.dev20260803+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.10.20 | packaged by conda-forge | (main, Mar  5 2026, 16:42:22) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-134-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      42%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.77
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.14.0.dev20260803+cpu
[pip3] torchvision==0.29.0.dev20260804+cpu
[pip3] triton==3.6.0
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] nvidia-cublas             13.1.0.3                 pypi_0    pypi
[conda] nvidia-cuda-cupti         13.0.85                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc         13.0.88                  pypi_0    pypi
[conda] nvidia-cuda-runtime       13.0.96                  pypi_0    pypi
[conda] nvidia-cudnn-cu13         9.19.0.56                pypi_0    pypi
[conda] nvidia-cufft              12.0.0.61                pypi_0    pypi
[conda] nvidia-curand             10.4.0.35                pypi_0    pypi
[conda] nvidia-cusolver           12.0.4.66                pypi_0    pypi
[conda] nvidia-cusparse           12.6.3.3                 pypi_0    pypi
[conda] nvidia-cusparselt-cu13    0.8.0                    pypi_0    pypi
[conda] nvidia-nccl-cu13          2.28.9                   pypi_0    pypi
[conda] nvidia-nvjitlink          13.0.88                  pypi_0    pypi
[conda] nvidia-nvtx               13.0.85                  pypi_0    pypi
[conda] torch                     2.14.0.dev20260803+cpu          pypi_0    pypi
[conda] torchvision               0.29.0.dev20260804+cpu          pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @jianyuh @nikitaved @mruberry @walterddr @Lezcano @Chillee @samdow @kshitij12345

**Comment by Skylion007** [comment]
PRs welcome here

**Comment by Saad-Mallebhari** [comment]
Hi @shadow-orange41 @Skylion007,

I investigated this nested forward-mode AD behavior.

###  Cause 
For $f(A) = \det(A)$, the forward-mode JVP rule evaluates $\text{JVP} = \det(A) \cdot \text{Tr}(A^{-1} V)$.

When computing second-order forward derivatives (`jacfwd(jacfwd)` or `jvp(jvp)`), the nested tangent propagation differentiates through the intermediate $A^{-1}$ term (which requires $\text{d}(A^{-1}) = -A^{-1} (\text{d}A) A^{-1}$). In the current JVP rule for `linalg.det` / `slogdet` / `linalg.solve`, tangent propagation through the intermediate inverse/solve step is either using an un-differentiable forward-AD path or failing to chain the matrix inverse tangent correctly.

### Fix Plan
1. Update the forward-mode AD (JVP) rules for `linalg.det`, `slogdet`, and `linalg.solve` to ensure tangent computation through $A^{-1}$ is fully differentiable under nested `jvp`.
2. Add regression unit tests in `test_autograd.py` / `test_functorch.py` covering second-order `jvp(jvp)` and `jacfwd(jacfwd)` for `linalg.det`, `slogdet`, and `linalg.solve`.

I am working on this fix and will submit a PR shortly!

**Comment by HussainNizamani** [comment]
Reproduced and root-caused. First, @Saad-Mallebhari: apologies for the timing collision. I've been on this since yesterday (reproduced on both nightly and main, root cause traced through the saved-LU path in `linalg_det_jvp`) but hadn't posted here yet, and that's on me. I have a complete fix with regression tests, validated to machine precision across all six second-order compositions for det, slogdet, and solve,
also Your cause analysis matches what I found. If you've already built something, and  i'm happy to compare approaches or co-review.

**can i open a draft PR today.** @Skylion007 


The detail below comes from my AI-assisted analysis (Claude, working under my direction); I've verified all of it on my own hardware.

> **Root cause.** The forward-AD formulas for `_linalg_det`, `_linalg_slogdet` and `_linalg_solve_ex` (`linalg_det_jvp` and friends in `torch/csrc/autograd/FunctionsManual.cpp`) compute their `A⁻¹·(...)` terms with `at::linalg_lu_solve` on the LU factors saved from the primal computation, and those auxiliaries are marked non-differentiable in `derivatives.yaml`. The helper's own comment says "Since we never differentiate over forward AD, we don't need to deal with further gradients". So when the JVP is differentiated again (`jvp∘jvp`, `grad∘jvp`, `jacfwd∘jacfwd`, `jacrev∘jacfwd`), the dependence of `A⁻¹` on `A` is invisible to autograd and only the `det(A)` factor contributes. The result is a silently wrong second derivative, while first order and reverse-over-reverse stay correct (the backward already recomputes differentiably for higher-order use).
>
> **Why CI never caught it.** `gradgradcheck` supports `check_fwd_over_rev` and `check_rev_over_rev` only. There is no forward-over-forward mode anywhere in the framework, so no OpInfo flag could exercise a doubly-forward composition.
>
> **The fix.** Pass the primal `A` into the three JVP helpers and recompute via `at::linalg_solve(A, ...)` whenever the result can be differentiated again (`A.requires_grad()` or functorch/subclass tensors involved), mirroring the existing higher-order branch in `linalg_det_backward`; otherwise keep the saved-LU fast path. Being upfront about the cost: classic `autograd.forward_ad` keeps the fast path (verified by op-profiling), but any `torch.func.jvp`/`jacfwd` call takes the recompute path, since at this layer a functorch wrapper can't be told apart from a pending higher-order differentiation. That's one extra factorization, only under torch.func, only for these three ops. If someone knows a sharper gate that doesn't reach into functorch internals, I'd like to hear it.
>
> **Validation** (Linux aarch64, main): the 2x2 analytic Hessian check goes from `max|H - H_true| = 1.166e+00` on all four forward-containing modes to `2.22e-16` on all six; slogdet was off by 8.0e-01 and solve by 1.2e+00 before, both at machine precision after. New regression tests fail before the fix and pass after. `test_linalg` det/slogdet/solve subset: 99 passed. `functorch/test_ops` filtered: 141 passed. Zero failures.
>
> **Related: #192521 is a second, distinct defect in the same file.** Under torch.func, functorch-wrapped tensors count as tensor-subclass-like, and `linalg_det_backward`'s subclass branch unconditionally used the singular-case SVD-adjugate formula. `svd_backward`'s `1/(σᵢ² - σⱼ²)` terms then poison well-conditioned inputs whose singular values are merely close, and NaN on exact ties. The fix replaces that branch with a vmap-safe `where`-blend that evaluates each formula on inputs sanitized so both branches stay finite where unselected. The clustered-Q case and the exact-tie identity case both come out correct and finite, and the same PR closes both issues.

**Comment by lezcano** [comment]
@HussainNizamani sure do send that fix.

**Comment by HussainNizamani** [comment]
PR is up: #192667. It fixes both #192540 and #192521: the forward-AD formulas for det/slogdet/solve now receive the primal `A` through `derivatives.yaml` and recompute via `linalg_solve(A, ...)` when the JVP can be differentiated again, mirroring the existing higher-order branch of `linalg_det_backward`; and the tensor-subclass branch of `linalg_det_backward` becomes a sanitized `where`-blend so clustered or repeated singular values stay finite. 
Regression tests live in `test/test_linalg.py` (device-generic, seeded, including an analytic 2x2 Hessian oracle borrowed from #192596's idea) and fail before the fix.


*(AI-Assisted)*
